botocore/errorfactory.py

```diff
@@ -0,0 +1,42 @@
+import logging
+import threading
+
+from botocore.utils import CachedProperty
+from botocore.exceptions import ClientError
+
+logger = logging.getLogger(__name__)
+
+
+class ServiceErrorFactory:
+    ClientError = ClientError
+
+    def __init__(self, service_model):
+        self._service_model = service_model
+        self._lock = threading.Lock()
+
+    @CachedProperty
+    def _error_shapes(self):
+        shapes = {}
+        for op_name in self._service_model.operation_names:
+            op_model = self._service_model.operation_model(op_name)
+            for shape in op_model.error_shapes:
+                shapes[shape.name] = shape
+        return shapes
+
+    def __getattr__(self, attr):
+        if attr.startswith("_"):
+            raise AttributeError(attr)
+        if attr not in self._error_shapes:
+            msg = 'Error "{0}" was not found in the service model'
+            logger.warning(msg.format(attr))
+        with self._lock:
```

Comment:
The hypothetical concern here is a race condition where two threads both say `x = exceptions.X` and both factories decide to create a class for it. Now we have two different identities for the exception class X - one canonical, and one orphaned. The thread whose reference got overwritten will fail to catch the resulting exception with `try .. except X`, because it is no longer an instance of X.